src/Cooked/MockChain/GenerateTx/Mint.hs

```diff
@@ -7,24 +7,25 @@ import Cooked.MockChain.GenerateTx.Common
 import Cooked.MockChain.GenerateTx.Witness
 import Cooked.Skeleton
 import Data.Map qualified as Map
+import Data.Map.NonEmpty qualified as NEMap
+import Data.Map.Strict qualified as SMap
+import GHC.Exts (fromList)
 import Ledger.Tx.CardanoAPI qualified as Ledger
 import Plutus.Script.Utils.Scripts qualified as Script
+import PlutusLedgerApi.V3 qualified as Api
+import PlutusTx.Builtins.Internal qualified as PlutusTx
+import Test.QuickCheck.Modifiers (NonZero (NonZero))
 
 -- | Converts the 'TxSkelMints' into a 'TxMintValue'
 toMintValue :: (MonadBlockChainBalancing m) => TxSkelMints -> m (Cardano.TxMintValue Cardano.BuildTx Cardano.ConwayEra)
 toMintValue mints | null mints = return Cardano.TxMintNone
-toMintValue mints | mintValue <- txSkelMintsValue mints = do
-  mintVal <-
-    throwOnToCardanoError
-      ("toMintValue: Unable to translate minted value " <> show mintValue)
-      (Ledger.toCardanoValue mintValue)
-  (Map.fromList -> witnessMap) <-
-    forM (txSkelMintsToList mints) $
-      \(policy, redeemer, _, _) -> do
-        policyId <-
-          throwOnToCardanoError
-            "toMintValue: Unable to translate minting policy hash"
-            (Ledger.toCardanoPolicyId (Script.mintingPolicyHash policy))
-        mintWitness <- toScriptWitness policy redeemer Cardano.NoScriptDatumForMint
-        return (policyId, mintWitness)
-  return $ Cardano.TxMintValue Cardano.MaryEraOnwardsConway mintVal (Cardano.BuildTxWith witnessMap)
+toMintValue mints = fmap (Cardano.TxMintValue Cardano.MaryEraOnwardsConway . SMap.fromList) $
+  forM (Map.toList mints) $ \(policy, (red, assets)) -> do
+    policyId <-
+      throwOnToCardanoError
+        "toMintValue: Unable to translate minting policy hash"
+        (Ledger.toCardanoPolicyId $ Script.toMintingPolicyHash policy)
+    mintWitness <- toScriptWitness policy red Cardano.NoScriptDatumForMint
+    assetsMinted <- forM (Map.toList $ NEMap.toMap assets) $ \(Api.TokenName (PlutusTx.BuiltinByteString name), NonZero quantity) -> do
+      return (Cardano.AssetName name, Cardano.Quantity quantity)
```

Comment:
My Haskell fu might be missing here, but why is there a `do` here (end of line) immediately followed by `return`? isn't that equivalent to just getting rid of the `do` and using `return`/`pure` directly?

Reply:
This is correct thanks ! Probably a remnant of a former longer block !
Actually it's not even a monadic call so I likely can further improve this code and change the `forM` into a regular `<$>` !